docs/operating-scylla/procedures/cluster-management/handling-membership-change-failures.rst

```diff
@@ -0,0 +1,204 @@
+Handling Cluster Membership Change Failures
+*******************************************
+
+A failure may happen in the middle of a cluster membership change (that is bootstrap, decommission, removenode, or replace), such as loss of power. If that happens, you should ensure that the cluster is brought back to a consistent state as soon as possible. Further membership changes might be impossible until you do so.
+
+For example, a node that crashed in the middle of decommission might leave the cluster in a state where it considers the node to still be a member, but the node itself will refuse to restart and communicate with the cluster. This particular case is very unlikely - it requires a specifically timed crash to happen, after the data streaming phase of decommission finishes but before the node commits that it left. But if it happens, you won't be able to bootstrap other nodes (they will try to contact the partially-decommissioned node and fail) until you remove the remains of the node that crashed.
+
+---------------------------
+Handling a Failed Bootstrap
+---------------------------
+
+If a failure happens when trying to bootstrap a new node to the cluster, you can try bootstrapping the node again by restarting it.
+
+If the failure persists or you decided that you don't want to bootstrap the node anymore, follow the instructions in the :ref:`cleaning up after a failed membership change <cleaning-up-after-change>` section to remove the remains of the bootstrapping node. You can then clear the node's data directories and attempt to bootstrap it again.
+
+------------------------------
+Handling a Failed Decommission
+------------------------------
+
+There are two cases.
+
+Most likely the failure happened during the data repair/streaming phase - before the node tried to leave the token ring. Look for a log message containing "leaving token ring" in the logs of the node that you tried to decommission. For example:
+
+.. code-block:: console
+
+    INFO  2023-03-14 13:08:38,323 [shard 0] storage_service - decommission[5b2e752e-964d-4f36-871f-254491f4e8cc]: leaving token ring
+
+If the message is **not** present, the failure happened before the node tried to leave the token ring. In that case you can simply restart the node and attempt to decommission it again.
+
+If the message is present, the node attempted to leave the token ring, but it might have left the cluster only partially before the failure. **Do not try to restart the node**. Instead, you must make sure that the node is dead and remove any leftovers using the :doc:`removenode operation </operating-scylla/nodetool-commands/removenode/>`. See :ref:`cleaning up after a failed membership change <cleaning-up-after-change>`. Trying to restart the node after such failure results in unpredictable behavior - it may restart normally, it may refuse to restart, or it may even try to rebootstrap.
```

Comment:
This looks a bit scary and gives an impression that the system is fragile. If we tell users to not attempt the restart after certain thing is logged, the node could do this automatically and record a flag locally and refuse to boot if the flag is set, and tell the user what to do.

Reply:
> This looks a bit scary and gives an impression that the system is fragile.

Because it is.

Fixing the fragility > documenting the fragility > doing nothing.

I support fixing the fragility eventually, but instead of constantly patching the legacy gossiper-based topology changes code which will always be fragile in some ways, we should focus on delivering consistent raft-based topology changes with fencing and all that. The legacy can stay documented and eventually we'll disable it.

> If we tell users to not attempt the restart after certain thing is logged, the node could do this automatically and record a flag locally and refuse to boot if the flag is set, and tell the user what to do.

In this particular scenario, the node may try to rebootstrap after leaving token ring precisely _because_ it's setting a flag:
```cpp
future<> storage_service::leave_ring() {
    co_await _sys_ks.local().set_bootstrap_state(db::system_keyspace::bootstrap_state::NEEDS_BOOTSTRAP);
```
we have this code since 2015. Is there really a point in fixing it now?

Reply:
> In this particular scenario, the node may try to rebootstrap after leaving token ring precisely because it's setting a flag:

I meant a flag which is set before the noreturn point. If that fails, we haven't reached the noreturn point yet, so can retry. If it succeeds, we're safe against restarts.

This is not a blocker for documenting it.